src/Common/src/System/Security/Cryptography/EccKeyFormatHelper.cs

```diff
@@ -254,15 +262,171 @@ private static ECCurve GetCurve(in ECDomainParameters domainParameters)
+                                k3 <= k2 ||
+                                k3 >= m)
+                            {
+                                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                            }
+
+                            pentanomialReader.ThrowIfNotEmpty();
+
+                            break;
+                        default:
+                            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
+
+                    innerReader.ThrowIfNotEmpty();
+
+                    BitArray poly = new BitArray(m + 1);
+                    poly.Set(m, true);
+                    poly.Set(k1, true);
+                    poly.Set(0, true);
+
+                    if (k2 > 0)
+                    {
+                        poly.Set(k2, true);
+                        poly.Set(k3, true);
+                    }
+
+                    primeOrPoly = new byte[(m + 7) / 8];
+                    poly.CopyTo(primeOrPoly, 0);
+                    Array.Reverse(primeOrPoly);
+                    break;
+                default:
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
             }
 
+            ECCurve curve;
+
+            if (prime)
+            {
+                curve = new ECCurve
+                {
+                    CurveType = ECCurve.ECCurveType.PrimeShortWeierstrass,
+                    Prime = primeOrPoly,
+                };
+            }
+            else
+            {
+                curve = new ECCurve
+                {
+                    CurveType = ECCurve.ECCurveType.Characteristic2,
+                    Polynomial = primeOrPoly,
+                };
+            }
+
+            curve.A = specifiedParameters.Curve.A.ToUnsignedIntegerBytes(primeOrPoly.Length);
+            curve.B = specifiedParameters.Curve.B.ToUnsignedIntegerBytes(primeOrPoly.Length);
+            curve.Order = specifiedParameters.Order.ToUnsignedIntegerBytes(primeOrPoly.Length);
+
+            ReadOnlySpan<byte> baseSpan = specifiedParameters.Base.Span;
+
+            // We only understand the uncompressed point encoding, but that's almost always what's used.
+            if (baseSpan[0] != 0x04 || baseSpan.Length != 2 * primeOrPoly.Length + 1)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            curve.G.X = baseSpan.Slice(1, primeOrPoly.Length).ToArray();
+            curve.G.Y = baseSpan.Slice(1 + primeOrPoly.Length).ToArray();
+
+            if (specifiedParameters.Cofactor.HasValue)
+            {
+                curve.Cofactor = specifiedParameters.Cofactor.Value.ToUnsignedIntegerBytes();
+            }
+
+            return curve;
+        }
+
+        internal static AsnWriter WriteSubjectPublicKeyInfo(in ECParameters ecParameters)
+        {
+            ecParameters.Validate();
```

Comment:
This (and all uses of _ecParameters_ later in this method) will make a copy of the struct since `ECParameters` and `Validate` are not marked readonly. Better to take the argument by value instead of by reference.